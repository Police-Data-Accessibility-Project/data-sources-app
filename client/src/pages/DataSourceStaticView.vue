<template>
	<div v-if="!noData">
		<FlexContainer component="header" alignment="center" class="data-details-header">
			<h2>{{ dataSource.name }}</h2>
			<Button>Edit</Button>
		</FlexContainer>
		<FlexContainer alignment="center" class="data-details-container">
			<div v-for="(section, index) in dataToRender" :key="index" class="data-detail-section">
				<h2>{{ section.header }}</h2>
				<div v-for="(record, recordIndex) in section.records" :key="recordIndex">
					<p class="large">{{ record.title }}</p>
					<div v-if="Array.isArray(dataSource[record.key])">
						<p v-for="item in dataSource[record.key]" :key="item" class="small">{{ item }}</p>
					</div>
					<p class="small" v-else>
						{{ dataSource[record.key] || 'null' }}
					</p>
				</div>
			</div>
		</FlexContainer>
	</div>
	<div v-else>
		<h2>{{ errorMessage }}</h2>
	</div>
</template>

<script>
import { Button, FlexContainer } from 'pdap-design-system';
import axios from 'axios';

export default {
	name: 'DataSourceStaticView',
	components: {
		Button,
		FlexContainer,
	},
	data: () => ({
		dataSource: {},
		id: null,
		dataToRender: [
			{
				header: 'Data Type',
				records: [
					{ title: 'Record type', key: 'record_type' },
					{ title: 'Description', key: 'description' },
					{ title: 'Tags', key: 'tags' },
				],
			},
			{
				header: 'Agency',
				records: [
					{ title: 'Name', key: 'agency_name' },
					{ title: 'State', key: 'state_iso' },
					{ title: 'County', key: 'county_name' },
					{ title: 'Municipality', key: 'municipality' },
					{ title: 'Agency Type', key: 'agency_type' },
					{ title: 'Jurisdiction Type', key: 'jurisdiction_type' },
				],
			},
			{
				header: 'Access & format',
				records: [
					{ title: 'Source URL', key: 'source_url' },
					{ title: 'ReadMe URL', key: 'readme_url' },
					{ title: 'Access Type', key: 'access_type' },
					{ title: 'Record Formats', key: 'record_format' },
					{ title: 'Detail Level', key: 'detail_level' },
					{ title: 'Size', key: 'size' },
					{ title: 'Access Notes', key: 'access_notes' },
					{ title: 'Records Not Online', key: 'records_not_online' },
				],
			},
			{
				header: 'Provenace',
				records: [
					{ title: 'Agency Supplied', key: 'agency_supplied' },
					{ title: 'Supplying Entity', key: 'supplying_entity' },
					{ title: 'Agency Originated', key: 'agency_originated' },
					{ title: 'Originating Entity', key: 'originating_entity' },
				],
			},
			{
				header: 'Coverage & retention',
				records: [
					{ title: 'Coverage Start Date', key: 'coverage_start' },
					{ title: 'Coverage End Date', key: 'coverage_end' },
					{ title: 'Source Last Updated', key: 'source_last_updated' },
					{ title: 'Update Frequency', key: 'update_frequency' },
					{ title: 'Update Method', key: 'update_method' },
					{ title: 'Retention Schedule', key: 'retention_schedule' },
					{ title: 'Number of Records Available', key: 'number_of_records_available' },
				],
			},
			{
				header: 'Data Source Meta',
				records: [
					{ title: 'Scraper URL', key: 'scraper_url' },
					{ title: 'Created', key: 'data_source_created' },
					{ title: 'Agency ID', key: 'agency_id' },
					{ title: 'Data Source ID', key: 'data_source_id' },
				],
			},
		],
		noData: true,
		errorMessage: '',
	}),
	mounted: function () {
		this.id = this.$route.params.id;
		this.getDataSourceDetails();
	},
	methods: {
		async getDataSourceDetails() {
			const headers = { Authorization: `Bearer ${process.env.VUE_APP_PDAP_TOKEN}` };
			try {
				const res = await axios.get(
					`${process.env.VUE_APP_BASE_URL}/search-tokens?endpoint=data-sources-by-id&arg1=${this.id}`,
					{ headers }
				);
				this.dataSource = res.data;
				this.noData = false;
			} catch (error) {
				this.errorMessage = error.response.data;
			}
		},
	},
};
</script>

<style scoped>
.data-details-header {
	justify-content: space-between;
	margin: 0 1rem;
}

.data-details-header .pdap-button {
	height: 50%;
}

.data-details-container {
	flex-flow: row wrap;
	align-items: stretch;
	gap: 2%;
	margin: 4rem 0;
}

.data-detail-section h2 {
	position: absolute;
	top: -1.75em;
	z-index: 1;
	padding: 0 0.5em;
	background-color: #fffbfa;
}

.data-detail-section {
	border: #000 1px solid;
	width: 45%;
	padding: 1%;
	position: relative;
	margin: 1% 0;
}

@media (prefers-color-scheme: dark) {
	.data-detail-section {
		border: #fffbfa 1px solid;
	}

	.data-detail-section h2 {
		background-color: #000;
	}
}
</style>
