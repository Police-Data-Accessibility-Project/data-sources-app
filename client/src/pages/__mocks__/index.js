export const resultsMock = {
	count: 3,
	data: [
		{
			agency_name: 'Cicero Police Department - IN',
			agency_supplied: false,
			airtable_uid: 'rec7edsmcuVQwMMXG',
			coverage_end: null,
			coverage_start: '2017-01-01',
			data_source_name: 'Calls for Service for Cicero Police Department - IN',
			description: null,
			municipality: 'Bridgeview',
			name: 'Cicero Police Department - IN',
			record_format: '[]',
			record_type: 'Calls for Service',
			source_url:
				'https://cityprotect.com/agency/cc0d9a3d-50b2-4424-ae25-5699ba6eaff9',
			state_iso: 'IN',
		},
		{
			agency_name: 'Chicago Police Department - IL',
			agency_supplied: true,
			airtable_uid: 'recv9fMNEQTbVarj2',
			coverage_end: null,
			coverage_start: '2019-01-01',
			data_source_name: 'Calls for Service for Chicago Police Department - IL',
			description: null,
			municipality: 'Chicago',
			name: 'Chicago Police Department - IL',
			record_format: '[]',
			record_type: 'Calls for Service',
			source_url:
				'https://informationportal.igchicago.org/911-calls-for-cpd-service/',
			state_iso: 'IL',
		},
		{
			agency_name: 'Chicago Police Department - IL',
			agency_supplied: false,
			airtable_uid: 'recv9fMNEQTbVarj2',
			coverage_end: null,
			coverage_start: '2018-12-18',
			data_source_name: '311 Calls for City of Chicago',
			description:
				'311 Service Requests received by the City of Chicago. This dataset includes requests created after the launch of the new 311 system on 12/18/2018 and some records from the previous system, indicated in the LEGACY\\_RECORD column.\n\nIncluded as a Data Source because in some cities 311 calls lead to police response; that does not appear to be the case in Chicago.\n',
			municipality: 'Chicago',
			name: 'Chicago Police Department - IL',
			record_format: "['CSV', 'XML', 'RDF', 'RSS']",
			record_type: 'Calls for Service',
			source_url:
				'https://data.cityofchicago.org/Service-Requests/311-Service-Requests/v6vf-nfxy',
			state_iso: 'IL',
		},
	],
};

export const dataSourceMock = {
	name: 'Calls for Service for Chicago Police Department - IL',
	submitted_name: 'Chicago Police Department',
	description: null,
	record_type: 'Calls for Service',
	source_url:
		'https://informationportal.igchicago.org/911-calls-for-cpd-service/',
	last_cached: '01/03/2024',
	agency_supplied: true,
	supplying_entity: null,
	agency_originated: null,
	originating_entity: null,
	agency_aggregation: null,
	coverage_start: '2019-01-01',
	coverage_end: null,
	source_last_updated: null,
	retention_schedule: null,
	detail_level: 'Summarized totals',
	number_of_records_available: null,
	size: null,
	access_type: ['Web page'],
	data_portal_type: null,
	record_format: null,
	update_frequency: 'Bi-weekly',
	update_method: null,
	tags: null,
	readme_url: null,
	scraper_url: null,
	data_source_created: '2022-09-28',
	airtable_source_last_modified: '2023-11-08',
	url_status: 'ok',
	rejection_note: null,
	last_approval_editor:
		'{"id": "usrtLIB4Vr3jTH8Ro", "email": "josh.chamberlain@pdap.io", "name": "Josh Chamberlain"}',
	agency_described_submitted: 'Chicago city, Illinois',
	agency_described_not_in_database: null,
	approval_status: 'approved',
	record_type_other: null,
	data_portal_type_other: null,
	records_not_online: null,
	data_source_request: null,
	url_button: null,
	tags_other: null,
	access_notes: null,
	homepage_url: 'https://home.chicagopolice.org/',
	count_data_sources: 21,
	agency_type: 'law enforcement/police',
	multi_agency: null,
	jurisdiction_type: 'local',
	state_iso: 'IL',
	municipality: 'Chicago',
	zip_code: '60616',
	county_fips: '17031',
	county_name: ['Cook'],
	lat: 41.85861,
	lng: -87.62802,
	data_sources: [
		'recs7gYAGAMWEhLYC',
		'recyR4IVgqyIekzYB',
		'recBMxu4UcHCqjsWQ',
		'recqXzWtmhfV9z8Av',
		'recc5QWeCMGL17Ab6',
		'recD9bNxneanhBBDH',
		'recPUGM4OEopoRAXB',
		'rec10Mts7CmMPzcnx',
		'recTj2zTdu7Jp8oX9',
		'rec00T2YLS2jU7Tbn',
		'recpOsfsTV4bKmVfa',
		'recipwTdMN0jRADrt',
		'recjFk7X0KofjTPGK',
		'rec0Lqwrj6blSbkuz',
		'recYyAY8XVQJJVde6',
		'recgwPuRTwjAMUZeO',
		'recOLt1yQGlgrO5jQ',
		'recdAte1fo5FBruE7',
		'recUGIoPQbJ6laBmr',
		'recTw75rItd827L0r',
		'rechI06qD4od759xT',
		'recodGxvuxiTXPnOL',
	],
	no_web_presence: null,
	airtable_agency_last_modified: '2023-05-16',
	data_sources_last_updated: '2023-04-06',
	approved: true,
	rejection_reason: null,
	agency_created: '2022-08-18',
	county_airtable_uid: 'recuhY2ud60V41j0w',
	defunct_year: null,
	data_source_id: 'rec00T2YLS2jU7Tbn',
	agency_id: 'recv9fMNEQTbVarj2',
	agency_name: 'Chicago Police Department - IL',
};
