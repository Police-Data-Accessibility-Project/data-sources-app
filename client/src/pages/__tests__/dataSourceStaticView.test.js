import { flushPromises, mount } from '@vue/test-utils';
import DataSourceStaticView from '../DataSourceStaticView.vue';
import { beforeAll, beforeEach, describe, expect, it, vi } from 'vitest';
import axios from 'axios';
import { nextTick } from 'vue';
import { dataSourceMock } from '../__mocks__';

let wrapper;

const push = vi.fn();
const $routerMock = {
	push,
};
const $routeMock = {
	params: {
		id: 'test',
	},
};
vi.mock('axios');

beforeAll(() => {
	import.meta.env.VITE_VUE_APP_BASE_URL = 'https://data-sources.pdap.io';
});

describe('DataSourceStaticView', () => {
	beforeEach(() => {
		wrapper = mount(DataSourceStaticView, {
			global: {
				mocks: {
					$router: $routerMock,
					$route: $routeMock,
				},
			},
		});

		axios.get.mockResolvedValue({ data: dataSourceMock });
	});

	it('Calls API and renders search results', async () => {
		const id = wrapper.vm.id;

		await wrapper.vm.getDataSourceDetails();
		await flushPromises();

		// TODO: figure out why this is called twice
		expect(axios.get).toHaveBeenCalledTimes(2);
		expect(axios.get).toHaveBeenLastCalledWith(
			`https://data-sources.pdap.io/search-tokens?endpoint=data-sources-by-id&arg1=${id}`,
		);

		await nextTick();

		expect(wrapper.find('[data-test="data-source-static-view"]').exists()).toBe(
			true,
		);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('Routes back to / on new search button click', () => {
		wrapper.find('[data-test="new-search-button"]').trigger('click');
		expect(push).toHaveBeenCalledWith('/');
	});

	it('Routes back to /search on Record Type button click with correct parameters', async () => {
		const button = wrapper.find('[data-test="record-type-button"]');

		expect(button.exists()).toBe(true);

		const type = button.text();

		button.trigger('click');

		await nextTick();

		expect(push).toHaveBeenLastCalledWith(`/search/${type}/all`);
	});

	it('Routes back to /search on Agency Name button click with correct parameters', async () => {
		const button = wrapper.find('[data-test="agency-name-button"]');

		expect(button.exists()).toBe(true);

		const name = button.text();

		button.trigger('click');

		await nextTick();

		expect(push).toHaveBeenLastCalledWith(`/search/all/${name}`);
	});

	it('Opens link on archive button click.', async () => {
		const spy = vi.spyOn(window, 'open');
		const button = wrapper.find('[data-test="view-archives-button"]');

		expect(button.exists()).toBe(true);

		button.trigger('click');

		await nextTick();

		expect(spy).toHaveBeenLastCalledWith(
			`https://web.archive.org/web/*/${dataSourceMock.source_url}`,
			'_blank',
		);
	});

	// it("renders correctly when there is no data", () => {
	// 	const wrapper = mount(DataSourceStaticView, {
	// 		data() {
	// 			return {
	// 				noData: true,
	// 				errorMessage: "No data available",
	// 			};
	// 		},
	// 	});

	// 	expect(wrapper.find("h2").text()).toBe("No data available");
	// });

	// Add more test cases as needed
});
