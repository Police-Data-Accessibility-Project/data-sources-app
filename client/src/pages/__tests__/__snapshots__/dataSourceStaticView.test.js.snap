// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataSourceStaticView > Calls API and renders search results 1`] = `
<main class="flex flex-col p-8 w-full">
  <h1 class="flex justify-start mt-2 w-full">Calls for Service for Chicago Police Department - IL</h1>
  <button class="pdap-button pdap-button-secondary new-search">
    <i class="fa fa-plus"></i> New search
  </button>
  <div alignment="center" class="grid grid-cols-1 w-full auto-rows-auto gap-4 mx-auto items-start md:grid-cols-2 lg:grid-cols-3">
    <!-- Each card -->
    <div class="flex self-stretch flex-col text-lg">
      <div class="border border-neutral-400 h-full leading-snug p-3 mt-8">
        <h2 class="text-xl font-semibold line-clamp-2">Data Type</h2><!-- Within each card, the items listed -->
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Record type</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <button class="pdap-button pdap-button-tertiary mt-1 py-[.125rem] px-3 rounded-full bg-brand-wine/10 dark:bg-brand-wine dark:text-white w-fit small font-normal text-lg" href="Calls for Service">Calls for Service</button>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Description</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Tags</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
      </div>
    </div>
    <div class="flex self-stretch flex-col text-lg">
      <div class="border border-neutral-400 h-full leading-snug p-3 mt-8">
        <h2 class="text-xl font-semibold line-clamp-2">Access & format</h2><!-- Within each card, the items listed -->
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Source URL</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <a class="w-full inline-block truncate-text" href="https://informationportal.igchicago.org/911-calls-for-cpd-service/" rel="_blank" target="_blank">https://informationportal.igchicago.org/911-calls-for-cpd-service/</a>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title"></p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, this isn't returned from API, we need to do it ourselves -->
          <pbutton class="flex gap-4 items-center" intent="secondary">View Archives <i class="fa fa-external-link"></i>
          </pbutton>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Last Archived</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="01/03/2024">01/03/2024</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">ReadMe URL</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Access Type</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <div>
            <p class href="Web page">Web page</p>
          </div>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Record Formats</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Detail Level</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="Summarized totals">Summarized totals</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Size</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Access Notes</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
      </div>
    </div>
    <div class="flex self-stretch flex-col text-lg">
      <div class="border border-neutral-400 h-full leading-snug p-3 mt-8">
        <h2 class="text-xl font-semibold line-clamp-2">Agency</h2><!-- Within each card, the items listed -->
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Name</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <button class="pdap-button pdap-button-tertiary decoration-solid decoration-[6%] text-brand-gold  underline-offset-[5%] hover:brightness-85 small p-0 font-normal text-lg" href="Chicago Police Department - IL">Chicago Police Department - IL</button>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">State</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="IL">IL</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">County</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <div>
            <p class href="Cook">Cook</p>
          </div>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Municipality</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="Chicago">Chicago</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Agency Type</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="law enforcement/police">law enforcement/police</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Jurisdiction Type</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="local">local</p>
        </div>
      </div>
    </div>
    <div class="flex self-stretch flex-col text-lg">
      <div class="border border-neutral-400 h-full leading-snug p-3 mt-8">
        <h2 class="text-xl font-semibold line-clamp-2">Provenance</h2><!-- Within each card, the items listed -->
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Agency Supplied</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="true">true</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Supplying Entity</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Agency Originated</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Originating Entity</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
      </div>
    </div>
    <div class="flex self-stretch flex-col text-lg">
      <div class="border border-neutral-400 h-full leading-snug p-3 mt-8">
        <h2 class="text-xl font-semibold line-clamp-2">Coverage & retention</h2><!-- Within each card, the items listed -->
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Coverage Start Date</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="2019-01-01">2018</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Coverage End Date</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Source Last Updated</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Update Frequency</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="Bi-weekly">Bi-weekly</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Update Method</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Retention Schedule</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Number of Records Available</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
      </div>
    </div>
    <div class="flex self-stretch flex-col text-lg">
      <div class="border border-neutral-400 h-full leading-snug p-3 mt-8">
        <h2 class="text-xl font-semibold line-clamp-2">Data Source Meta</h2><!-- Within each card, the items listed -->
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Scraper URL</p><!-- If no data for this key, em-dash -->
          <div class="text-neutral-500"> — </div><!-- If data is an array and returned from API, loop over it and render each -->
          <!--v-if-->
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Created</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="2022-09-28">09/27/2022</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Agency ID</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="recv9fMNEQTbVarj2">recv9fMNEQTbVarj2</p>
        </div>
        <div>
          <!-- Title -->
          <p class="large text-brand-wine dark:text-white font-semibold text-sm uppercase tracking-wider mb-0 mt-4" v:if="record.title">Data Source ID</p><!-- If no data for this key, em-dash -->
          <!--v-if-->
          <!-- If data is an array and returned from API, loop over it and render each -->
          <!-- Otherwise, if single item returned from API, render that item  -->
          <!-- TODO: Abstract this duplicate logic into a render function (https://vuejs.org/guide/extras/render-function) to decrease repetition  -->
          <p class href="rec00T2YLS2jU7Tbn">rec00T2YLS2jU7Tbn</p>
        </div>
      </div>
    </div>
  </div>
</main>
`;
